### DISCUSSION TITLE
How to share a context in server components

### DISCUSSION THREAD

**Discussion by Amerehei** [vercel/next.js#60520] [App Router]
In Next.JS app router we are dealing with a lot of files some of components are server components and some others are client components

For client components, When we need to share an object we have options such as passing props, using context and state managers like redux

In server components, especially when we are dealing with multiple files there is no way to share data

Consider we have a global `layout.tsx` which is a server component, in this component we fetch some data and calculated some information (link auth and user info). how can we share these data in another inner `layout.tsx` or server component `page.tsx`?

**Comment by gVguy** [comment]
I want to know this as well. It seems like there's no way to avoid prop drilling in server components.

You can see this problem in this [example app](https://github.com/vercel/next.js/blob/canary/examples/app-dir-i18n-routing/app/%5Blang%5D/page.tsx) from the official docs.

We need to pass the `dictionary` down as a prop to the nested component, because the `lang` param can only be accessed from the page component itself. Even in a small example project like the linked one, this looks ugly, imagine a large-scale application, it would be a nightmare to pass the dictionary to every single component responsible for rendering text.

There should be a way to provide values in server components to cover use cases like this one.

**Comment by don-esteban** [comment]
I use [React cache](https://react.dev/reference/react/cache) for similar use cases.

```
import { cache } from 'react';

export const getExpensiveData = cache(function getExpensiveData(): ExpensiveData {
// Retrieve data and do expensive processing
  return expensiveData;
})
```

For retrieval of external data without expensive calculations, I use [Next.js fetch](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating), which supports caching and other stuff.
